subsystem/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+        std::mem::drop(self.shutting_down_tx);
+
+        log::info!("Subsytem manager {} shutting down", self.name);
+
+        // Order all the remaining subsystems to shut down.
+        let _ = self.shutdown_request_tx.send(());
+
+        // Wait for the subsystems to go down.
+        // TODO add shutdown timeout here
+        Self::wait_for_subsystems_to_shut_down(self.shutting_down_rx).await;
+
+        log::info!("Subsystem manager {} terminated", self.name);
+    }
+}
+
+/// Internal action type sent in the channel.
+type Action<T> = Box<dyn FnOnce(&mut T) + Send>;
+
+/// Call request
+pub struct CallRequest<T>(mpsc::Receiver<Action<T>>);
+
+impl<T: 'static + Send> CallRequest<T> {
+    /// Receive an external call to this subsystem.
+    pub async fn recv(&mut self) -> Action<T> {
+        match self.0.recv().await {
+            // We have a call, return it
+            Some(action) => action,
+            // All handles to this subsystem dropped, suspend call handling.
+            None => std::future::pending().await,
+        }
+    }
+}
+
+/// Shutdown request
+pub struct ShutdownRequest(broadcast::Receiver<()>);
+
+impl ShutdownRequest {
+    /// Receive a shutdown request.
+    pub async fn recv(&mut self) {
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
+            }
+            Err(broadcast::error::RecvError::Closed) => {
+                panic!("Shutdown channel sender closed prematurely")
+            }
+            Ok(()) => (),
+        }
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+#[derive(Clone)]
+pub struct Subsystem<T> {
+    // Send the subsystem stuff to do.
+    action_tx: mpsc::Sender<Action<T>>,
+}
+
+impl<T: Send + 'static> Subsystem<T> {
+    /// Crate a new subsystem handle.
+    fn new(action_tx: mpsc::Sender<Action<T>>) -> Self {
+        Self { action_tx }
+    }
+
+    /// Dispatch a function call to the subsystem
+    pub async fn call_mut<R: Send + 'static>(
+        &self,
+        func: impl FnOnce(&mut T) -> R + Send + 'static,
+    ) -> R {
+        let (rtx, rrx) = oneshot::channel::<R>();
+
+        self.action_tx
+            .send(Box::new(move |subsys| {
+                let result = func(subsys);
+                rtx.send(result).ok().expect("Value return channel closed");
+            }))
+            .await
+            .ok()
+            .expect("Target subsystem down");
+
+        rrx.await.expect("Target subsystem down")
+    }
+
+    /// Dispatch a function call to the subsystem (convenience method if subsystem is not mutated)
+    pub async fn call<R: Send + 'static>(&self, func: impl FnOnce(&T) -> R + Send + 'static) -> R {
+        self.call_mut(|this| func(this)).await
```

Comment:
Right now they're both the same. Later, we need to allow read-only calls to be done in paralle, but read/write calls to be done uniquely. Just like an rw-lock. While this isn't a requirement right now, you're free to plan for it now.

Reply:
Maybe there's an angle I'm not seeing. But assuming that const-correctness is held, we can guarantee that no side effects will happen. At least that's one of my design goals.

Reply:
Looking at it now, I just realised: The subsystem is already in control what to pass into the function as the argument. I.e. in the code snippet `subsys.call(|this| this.foo(5))`, the subsystem decides what to pass in as `this`. Passing in a reference is a relic of the times when only passive subsystems were supported, where a mutable reference to the subsystem object was always passed. Currently, the subsystem can pass a mutable reference to a value of its own choosing. It makes sense to me to get rid of the reference and let the subsystem pass in a value of any type `T`. We can recover the original behaviour by setting `T = &mut MySubsystem`. This makes some parts of the code a bit less awkward and also removes the need for this distinction between `call` and `call_mut`.